lib/database/commands.rb

```diff
@@ -0,0 +1,880 @@
+      @_last_response = nil      
+    end
+
+    def server_response
+      @_last_response
+    end  
+
+    def was_failed?()
+      !@failed_nodes.empty?
+    end
+
+    def add_failed_node(node)
+      assert_node(node)
+      @failed_nodes.add(node)
+    end
+
+    def was_failed_with_node?(node)
+      assert_node(node)
+      @failed_nodes.include?(node)
+    end
+
+    def create_request(server_node)
+      raise NotImplementedError, "You should implement create_request method"
+    end  
+
+    def to_request_options
+      end_point = @end_point
+
+      if !@params.empty?        
+        encoded_params = URI.encode_www_form(@params)
+        end_point = "#{end_point}?#{encoded_params}"
+      end
+
+      requestCtor = Object.const_get("Net::HTTP::#{@method.capitalize}")
+      request = requestCtor.new(end_point)
+
+      if !@payload.nil? && !@payload.empty?
+        request.body = JSON.generate(@payload)
+        @headers['Content-Type'] = 'application/json'
+      end 
+      
+      if !@headers.empty?      
+        @headers.each do |header, value|
+          request.add_field(header, value)
+        end
+      end  
+
+      request
+    end  
+
+    def set_response(response)
+      @_last_response = response
+
+      if @_last_response
+        ExceptionsRaiser.try_raise_from(response)
+        return response.json
+      end   
+    end  
+
+    protected
+    def assert_node(node)
+      raise ArgumentError, "Argument \"node\" should be an instance of ServerNode" unless node.is_a? ServerNode
+    end
+
+    def add_params(param_or_params, value)      
+      new_params = param_or_params
+
+      if !new_params.is_a?(Hash)
+        new_params = Hash.new
+        new_params[param_or_params] = value
+      end    
+
+      @params = @params.merge(new_params)
+    end
+
+    def remove_params(param_or_params, *other_params)
+      remove = param_or_params
+
+      if !remove.is_a?(Array)
+        remove = [remove]
+      end  
+
+      if !other_params.empty?        
+        remove = remove.concat(other_params)
+      end
+
+      remove.each {|param| @params.delete(param)}
+    end  
+  end  
+
+  class BatchCommand < RavenCommand
+    def initialize(commands_array = [])
+      super("", Net::HTTP::Post::METHOD)
+      @commands_array = commands_array
+    end
+
+    def create_request(server_node)
+      commands = @commands_array
+      assert_node(server_node)
+
+      if !commands.all? { |data| data && data.is_a?(RavenCommandData) }
+        raise InvalidOperationException, "Not a valid command"
+      end
+
+      @end_point = "/databases/#{server_node.database}/bulk_docs"
+      @payload = {"Commands" => commands.map { |data| data.to_json }}
+    end
+
+    def set_response(response)
+      result = super(response)
+
+      if !response.body
+        raise InvalidOperationException, "Invalid response body received"
+      end
+
+      result["Results"]
+    end
+  end
+
+  class CreateDatabaseCommand < RavenCommand
```

Comment:
Shouldn't all of this be in dedicated files?

Reply:
no. due to the issues of require in Ruby.
there is no way to export all classes form separated files like it does in nodeJS
also there are no classes autoloading by namespace like in Java, C# or PHP
so the better way - is to export a batch of classes in single file (for include them by single require statement). So for example when we want to use operations we just do require ravendb/database/operations instead of require each operaiton e needed

Reply:
@a-serdukov-mobilunity  https://stackoverflow.com/questions/12035057/breaking-ruby-module-across-several-files ?

Reply:
as i remember i tried it bot something won't work properly. but after my tries i updated to 2.4 so i can re-try again